### PR TITLE
Migrate to typescript

### PR DESCRIPTION
**Pull request by s1na**
This PR migrates the library to typescript and updates the tooling (`tslint`, `prettier`) to be consistent with https://github.com/ethereumjs/rlp/pull/37.

Additionally, it implements the logic of [ethUtil.defineProperties](https://github.com/ethereumjs/ethereumjs-util/blob/master/docs/index.md#defineproperties), in the `Account` class, with the help of a new class `BinaryValue`.

### CODE REVIEW

**Comment by holgerd77** [comment]
What's your main rationale to move the logic of this ``defineProperties`` method here (not against it, just asking)? And do you think you have met all the functionality relevant here, since the original method contains lot more code?

**Comment by holgerd77** [comment]
Also, if you have better ideas for the API respectively see some bad design choices feel free to introduce if you're confident on it. I have been moving more in the direction during the last weeks that we actually should allow us to also make API changes if useful and don't necessarily have to stick on transferring one-to-one. Confident that we get this managed, and the occasion is good to get rid of some legacy stuff.

**Comment by holgerd77** [comment]
//cc @krzkaczor did you also already stumble upon this defineProperties method and have ideas on that?

**Comment by s1na** [comment]
@holgerd77 Well, in the beginning it was because `defineProperties` modifies the class, adds properties and methods during runtime which typescript doesn't detect during compilation and therefore raises errors. But generally I think even though `defineProperties` is a neat abstraction, it's less readable and type-safe than explicitly defining properties and validating them.

I'm not sure if the current implementation covers all the cases, I have to test further. But it's less code because `Account` doesn't need some of the code in `defineProperties`, and they're there for other libraries, e.g. [Transaction](https://github.com/ethereumjs/ethereumjs-tx/blob/master/index.js#L50).

Agree with what you said on API change. Maybe transition to typescript and a systematic upgrade of API can be undertaken under codename EthereumJS 3.0 :smile:

File: package.json [hunk 1]
```diff
@@ -2,16 +2,28 @@
   "name": "ethereumjs-account",
   "version": "2.0.5",
   "description": "Encoding, decoding and validation of Ethereum's Account schema",
-  "main": "index.js",
+  "main": "./dist/index.js",
+  "types": "./dist/index.d.ts",
+  "files": [
+    "dist/***"
```
Comment (krzkaczor) [inline package.json]: is this a valid glob? shouldn't it be `dist/**/*`?

**Comment by krzkaczor** [comment]
I think the previous version with `defineProperties` allowed to set any properties — no we are missing setters. Frankly, if we consider breaking API we should make this immutable entirely, currently, it's mixed since there are methods like `setCode`.

On the other hand, if we don't want to break public interface, for now, I think that we could define typescript fields AND still use `defineProperties` which would override behavior in runtime. I think it should work and maybe even it's a preferred way since boilerplate footprint is small. @s1na What do you think about this approach. I am not sure if this gonna work, I would need to play with it.

File: package.json (hunk 1 above)
Comment (s1na) [reply package.json]: Ah good catch :)

**Comment by s1na** [comment]
@krzkaczor sorry, the current implementation is still work in progress, and some stuff that should be there and you mentioned are missing.

The approach you mentioned might be possible, please let me know if it worked out if you got around to playing with it. If we stick with this for now, in my opinion for future explicit definition of properties and their types feels safer and more readable than runtime injection, even if it increases boilerplate a bit. I'm trying to reduce the boilerplate by defining types that handle some of the validation done in `defineProperties`.

**Comment by holgerd77** [comment]
@s1na Yes, I also think it's very well worth to get rid of this common ``defineProperties`` method and it's worth to add some boilerplate here. These strong ties of core object structure changing to a dependency is really super-implicit, error-prone and has already caused various problems in the past. If this goes along with some changes to the API so be it, as long as it's thought through, provides some future-stable ground and has some reasoning for the changes.

**Comment by krzkaczor** [comment]
@s1na check it out here: https://github.com/ethereumjs/ethereumjs-account/pull/28/files

I think the most important part here is the usage of `!` to trick silent compiler warnings. 

Treat it as WIP. If you think it's a good idea incorporate please into your PR (you can rewrite it I just did it quick and dirty so maybe there some mistakes)

**Comment by s1na** [comment]
@krzkaczor Thanks! Integrated your code.

@holgerd77 The PR should be now ready for review, I'd appreciate it if you could have a look.

Defining the `Trie` interface here is probably redundant, but I didn't want to leave all the types `any`, and `merkle-patricia-tree` has not been migrated to typescript yet. After it is migrated, we can import it here and use its defined type.

**Comment by holgerd77** [comment]
Won't make it to do a review on this today - will try tomorrow - everyone else is also invited!

**Comment by holgerd77** [comment]
@s1na What happened to the `BinaryValue` class, did you remove that again?

**Comment by s1na** [comment]
@holgerd77 Yeah, I incorporated Chris's approach to make this PR only for transitioning to typescript and avoid API changes. We can later try to come up with a solution that'd work across projects. I'll create an issue for discussion.

**Comment by holgerd77** [comment]
@s1na Yeah, makes sense.

These `TypeScript` PRs are really some beasts, with changes in close to every file and all along the build chain. We should really make sure that we do at least one manual integration test on every transition by injecting the distribution build (`npm pack`) in the `node_modules` folder of a library where the respective library is used to check how things behave in production.

I have done this with this PR and it produces the following error when injecting in the VM and run `npm run testStateByzantium`:

```shell
error: TypeError: Account is not a constructor
    at /Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/tests/util.js:358:19
    at /Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:3096:16
    at replenish (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:998:17)
    at /Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:1002:9
    at eachLimit$1 (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:3182:24)
    at Object.<anonymous> (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:1032:16)
    at Object.exports.setupPreConditions (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/tests/util.js:356:9)
    at /Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/tests/GeneralStateTestsRunner.js:60:16
    at /Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:3866:24
    at replenish (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:998:17)
    at /Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:1002:9
    at eachOfLimit (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:1027:24)
    at /Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:1032:16
    at _parallel (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:3865:5)
    at Object.series (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/node_modules/async/dist/async.js:4721:5)
    at runTestCase (/Users/hdrewes/Documents/DEV/EthereumJS/ethereumjs-vm/tests/GeneralStateTestsRunner.js:48:9)
```

Not sure why this is happening yet, apparently `new Account()` constructor is not working properly on the distribution build. Have done some [googling](https://www.google.com/search?q=typescript+is+not+a+constructor&oq=typescript+is+not+a+con&aqs=chrome.1.69i57j0l5.5030j0j7&sourceid=chrome&ie=UTF-8), points in some "order of things is important" direction but couldn't completely identify the root cause yet.

This is now the third (so 3/3 😄) `TypeScript` transition PR together with `RLP` and `ethereumjs-common` where we have passing tests but breaking builds. Quite some track record! 😄 😛

**Comment by krzkaczor** [comment]
@holgerd77 

>This is now the third (so 3/3 😄) TypeScript transition PR together with RLP and ethereumjs-common where we have passing tests but breaking builds. Quite some track record! 😄 😛

This deserves a medal or something :D 

I stumbled upon this tweet and I keep wondering if its a big deal: https://twitter.com/bterlson/status/1082750157303603200

Anyways, probably the export "shape" is different.

**Comment by s1na** [comment]
@holgerd77 @krzkaczor Haha :)))

Apparently it's due to differences in module systems... when something is exported by `export default`, in node it has to be imported by `require(...).default`. If we don't do default export (and do named exports instead), it'd have to be imported by `require(..).Account`. The only way I [found](https://www.typescriptlang.org/docs/handbook/modules.html) to keep ES5 projects working as before is by using `export = Account`, which means in typescript the special syntax `import Account = require(..)` should be used!

When searching for this, it seems that generally named exports are preferred, unless there's only one object being exported (in which case default export is preferred).

Should we make this breaking change now, or rather go with `export = Account` for now and then after other libraries (VM) have also transitioned to typescript use proper exports? Or maybe someone has a better approach?

Did the same happen with `Common` Holger?

Edit: By the way, I tried compiling with `esModuleInterop` to no effect, it seems to be helpful for when we want to import JS libs (that use old export default method) in typescript, not the other way around.

**Comment by krzkaczor** [comment]
I am a huge hater of `import Account = require(..)` syntax: it's weird, uncommon and for example, babel doesn't support it.

**Comment by holgerd77** [comment]
Getting more and more to the stance that we just should do major releases by default on Typescript transition releases. Then we've got somewhat more freedom, the big shift gets clearer and it's generally somewhat likely that the WILL be done breaking changes on most of the transitions.

Probably happen on Common too, couldn't test yet. This will be a major release anyhow now though, just have to remember to test and state this in the release notes if the case.

**Comment by holgerd77** [comment]
No clear opinion on default vs named here, maybe default, sounds a bit more fitting?

**Comment by s1na** [comment]
VM has already upgraded to ethereumjs-util v6.0.0 and when I was testing `Account` in VM, I got errors for sha3 and SHA3_... constants, so I also upgraded ethereumjs-util here. Now all VM tests pass (after changing imports to `require('ethereumjs-account').default`).

**Comment by holgerd77** [comment]
Documentation should also be updated along with this, have made this an extra issue which can be addressed after merging: https://github.com/ethereumjs/ethereumjs-account/issues/30

**Comment by holgerd77** [comment]
@s1na I am not sure what to think about this external dependency (`Trie`) interface definitions included here. Are you confident that this has no side effects? @krzkaczor what's your stance here?

**Comment by s1na** [comment]
@holgerd77 It shouldn't have any effects beyond typescript compilation. If you think it's an unnecessary complication I can replace them all with `any`.

File: package.json [hunk 2]
```diff
@@ -24,14 +36,26 @@
   },
   "homepage": "https://github.com/ethereumjs/ethereumjs-account#readme",
   "dependencies": {
-    "ethereumjs-util": "^5.0.0",
-    "rlp": "^2.0.0",
+    "bn.js": "^4.11.8",
```
Comment (holgerd77) [inline package.json]: Where is this new `bn.js` dependency coming from respectively used?

File: package.json [hunk 3]
```diff
@@ -24,14 +36,26 @@
...
+    "ethereumjs-util": "^6.0.0",
+    "rlp": "^2.2.1",
     "safe-buffer": "^5.1.1"
   },
   "devDependencies": {
+    "@ethereumjs/config-nyc": "^1.0.0",
+    "@ethereumjs/config-prettier": "^1.0.1",
+    "@ethereumjs/config-tsc": "^1.0.2",
+    "@ethereumjs/config-tslint": "^1.0.0",
+    "@types/bn.js": "^4.11.3",
```
Comment (holgerd77) [inline package.json:48]: Same for the types?

**Comment by holgerd77** [comment]
@s1na If you are confident on that and @krzkaczor is not stating anything else leave them in, definitely nice to have if side-effect free.

File: package.json (hunk 3 above)
Comment (s1na) [reply package.json:48]: I'm not sure why, but without this typescript fails with:

```
node_modules/rlp/dist/types.d.ts:2:21 - error TS7016: Could not find a declaration file for module 'bn.js'. '.../ethereumjs-account/node_modules/bn.js/lib/bn.js' implicitly has an 'any' type.
  Try `npm install @types/bn.js` if it exists or add a new declaration (.d.ts) file containing `declare module 'bn.js';`

2 import BN = require('bn.js');
```

File: package.json (hunk 2 above)
Comment (s1na) [reply package.json]: I had mistakenly thought this is also necessary for the error (see below). Removed it now.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        length: 32,
+        default: ethUtil.KECCAK256_NULL,
+      },
+    ]
+
+    ethUtil.defineProperties(this, fields, data)
+  }
+
+  serialize(): Buffer {
+    return rlp.encode([this.nonce, this.balance, this.stateRoot, this.codeHash])
```
Comment (holgerd77) [inline src/index.ts:54]: Nice that you took out the ``raw`` here, much better readable like this.

**Review by holgerd77 — APPROVED** [review]
Rest looks good, thanks!

Yay. ✨ 😄

**Pull request merged**